stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/composer/MessageComposerView.kt

```diff
@@ -0,0 +1,195 @@
+     */
+    public var onCancelAction: () -> Unit = {}
+
+    /**
+     * @property onInputTextChanged Handler invoked when the user enters text in message input field.
+     */
+    public var onInputTextChanged: (String) -> Unit = {}
+
+    private val defaultLeadingContent: View by lazy {
+        val binding = StreamUiMessageComposerDefaultLeadingContentBinding.inflate(
+            streamThemeInflater,
+            binding.trailingContent,
+            false
+        )
+        binding.attachmentsButton.isVisible = false
+        binding.commandsButton.isVisible = false
+        binding.root
+    }
+
+    /**
+     * @property leadingContent A function returning content visible on the left.
+     */
+    public var leadingContent: MessageComposerView.() -> View = {
+        defaultLeadingContent
+    }
+        set(value) {
+            field = value
+            binding.leadingContent.apply {
+                removeAllViews()
+                addView(value())
+            }
+        }
+
+    private val defaultCenterContent: View by lazy {
+        val binding = StreamUiMessageComposerDefaultCenterContentBinding.inflate(
+            streamThemeInflater,
+            binding.trailingContent,
+            false
+        )
+        binding.messageEditText.addTextChangedListener(object : TextWatcherAdapter() {
+            override fun afterTextChanged(s: Editable) {
+                onInputTextChanged(s.toString())
+            }
+        })
+        binding.clearCommandButton.isVisible = false
+        binding.selectedAttachmentsContainer.isVisible = false
+        binding.commandBadge.isVisible = false
+        binding.messageReplyView.isVisible = false
+        CoroutineScope(Dispatchers.Main).launch {
+            messageInputState.collect {
+                binding.messageEditText.apply {
+                    val currentValue = text.toString()
+                    val newValue = it.inputValue
+                    if (newValue != currentValue) {
+                        setText(it.inputValue)
+                    }
+                }
+                binding.clearCommandButton.isVisible = it.inputValue.isNotEmpty()
+            }
+        }
+        binding.root
+    }
+
+    /**
+     * @property centerContent A function returning content visible in the middle.
+     */
+    public var centerContent: MessageComposerView.() -> View = {
+        defaultCenterContent
+    }
+        set(value) {
+            field = value
+            binding.centerContent.apply {
+                removeAllViews()
+                addView(value())
+            }
+        }
+
+    private val defaultTrailingContent: View by lazy {
+        val binding = StreamUiMessageComposerDefaultTrailingContentBinding.inflate(
+            streamThemeInflater,
+            binding.trailingContent,
+            false
+        ).apply {
+            CoroutineScope(Dispatchers.Main).launch {
```

Comment:
Same here - we're launching multiple scopes and coroutines in the Main thread to collect the same state - would it be easier to just collect all the state from the ViewModel in one place and update all the areas which care about the state?

Reply:
The motivation here was to allow user access to `messageInputState` in the scope of `MessageComposerView` when they override `MessageComposerView.trailingContent` for example. But maybe we could force them to subscribe to ViewModel and bind the `trailingContent: View` on their own and finally pass it to `MessageComposer` 🤔